File: 1115-print-foobar-alternately.py
class FooBar:
    def __init__(self, n):
        self.n = n
        self.cv = threading.Condition()
        self.is_foo = True

    def foo(self, printFoo: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            with self.cv:
                
                self.cv.wait_for(lambda : self.is_foo)
                
                # printFoo() outputs "foo". Do not change or remove this line.
                printFoo()
                
                self.is_foo = False
                self.cv.notify()


    def bar(self, printBar: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            with self.cv:
                self.cv.wait_for(lambda : not self.is_foo)
                
                # printBar() outputs "bar". Do not change or remove this line.
                printBar()
                
                self.is_foo = True
                self.cv.notify()

File: 1115-print-foobar-alternately.cpp
class FooBar {
private:
    int n;
    std::mutex m;
    std::condition_variable cv;
    bool is_foo;

public:
    FooBar(int n) {
        this->n = n;
        this->is_foo = true;
    }

    void foo(function<void()> printFoo) {
        
        for (int i = 0; i < n; i++) {
            
            std::unique_lock<std::mutex> lock(m);
            while (!is_foo)
            {
                cv.wait(lock);
            }

        	// printFoo() outputs "foo". Do not change or remove this line.
        	printFoo();
    
            is_foo = false;
            cv.notify_all();
        }
    }

    void bar(function<void()> printBar) {
        
        for (int i = 0; i < n; i++) {
            
            std::unique_lock<std::mutex> lock(m);
            while (is_foo)
            {
                cv.wait(lock);
            }

        	// printBar() outputs "bar". Do not change or remove this line.
        	printBar();
            
            is_foo = true;
            cv.notify_all();
        }
    }
};